### ISSUE TITLE
Allow for Dynamic Offset Searching in Auto Split state block

### ISSUE THREAD

**Issue by hawkerm** [LiveSplit/LiveSplit#487]
I'm writing an autosplitter for an old dos game which gets loaded in DOSBox.  I have found fixed offsets for the things I need from a fixed string within the game's memory.  However, the game's code is loaded at a different offset each time relative to DOSBox's address space.

I tried using SignatureScanner to look for my address, but it would only look so far ahead before failing:

    //WORKS vars.ccsaddr2 = new SignatureScanner(game, (IntPtr)vars.baddr + 0x4d5d000, 0xFFFFFF).Scan(new SigScanTarget(0, System.Text.Encoding.ASCII.GetBytes("STRINGHERE"))); 
    //FAILS vars.ccsaddr2 = new SignatureScanner(game, (IntPtr)vars.baddr, 0xFFFFFFF).Scan(new SigScanTarget(0, System.Text.Encoding.ASCII.GetBytes("STRINGHERE")));

I wrote a brute force loop to look for it which seemed to work until I found that my program could be in two different places 15mb apart.  It also means I have to rework all the old/current logic for all my values.

It'd be nice if there was a built in option where I could provide the search string identifier in the **state** block like it does currently for the alternate module, but have a way for it to mean it should go search and find that unique text within the memory space to work all other pointer path offsets in the state block so it would work with the current built-in old/current objects.

**Comment by fatalis** [comment]
I don't think this should be added, because it would require scanning the entire process' memory which could take over a minute for newer games. The reason SignatureScanner is failing for you is you're scanning invalid memory regions. You'd have to enumerate valid pages, which isn't possible with ASL due to it require P/Invoke. SignatureScanner is mostly meant for scanning the PE.

You should try to find a static base offset to build a pointer path. It may not be possible due to DOSBox being a VM though. Check the source code maybe?

**Comment by hawkerm** [comment]
The game executable is a compressed one, so it's unzipping the program into memory, so it's either DOSBox or that which is creating the dynamic offset (or a combination).

So, once the program is unzipped, I have all the fixed offsets, but I can't trace the whole invocation from the start of DOSBox to the uncompressing of the game to the game starting... it's never a fixed location (or set of locations), which is the problem.

I've seen two main possible 'fuzzy locations' so far, so if I look in the memory first and adjust my variable manually, I can get the following brute-force loop to work in a couple of seconds:

    int x = 0;
    while (memory.ReadString((IntPtr)vars.ccsaddr, 30) != "COPYRIGHT CODEMASTERS SOFTWARE" && x < 0x2000000)
    {
        x++;
        vars.ccsaddr = IntPtr.Add(vars.ccsaddr, 1);
    }

I think the address I'm looking for is at least block aligned, so I may be able to use that to skip through 16 bytes at a time, so I may be able to be crafty.

I tried looking at bigger chunks (below), but I'm guessing I ran into the paging problem, but it just would throw an 'object reference not set' exception without a line number, so I was in the dark.

    int x = 0;
    int y = -1;
    while (x < 0x2000000)
    {
        y = memory.ReadString((IntPtr)vars.ccsaddr, 1000000).IndexOf("COPYRIGHT");
        print("" + y);
        if (y != -1)
        {
            if (memory.ReadString((IntPtr)vars.ccsaddr + y, 30) == "COPYRIGHT CODEMASTERS SOFTWARE")
            {
                vars.ccsaddr = IntPtr.Add(vars.ccsaddr, y);
                break;
            }
        }
        vars.ccsaddr = IntPtr.Add(vars.ccsaddr, 999900);
        x += 999900;
    }

Is there some way I can detect that a memory read failed (most likely due to the paging issue)?

Thanks
